ortools/math_opt/solvers/xpress_solver.cc

```diff
@@ -0,0 +1,693 @@
+        << "Unable to enable XPRESS logs";
+  }
+  // Solve
+  if (is_mip_) {
+    ASSIGN_OR_RETURN(xpress_status_, xpress_->MipOptimizeAndGetStatus());
+  } else {
+    RETURN_IF_ERROR(SetLpIterLimits(parameters))
+        << "Could not set iteration limits.";
+    ASSIGN_OR_RETURN(xpress_status_, xpress_->LpOptimizeAndGetStatus(
+                                         GetOptimizationFlags(parameters)));
+  }
+  // Post-solve
+  if (!(is_mip_ ? (xpress_status_ == XPRS_MIP_OPTIMAL)
+                : (xpress_status_ == XPRS_LP_OPTIMAL))) {
+    RETURN_IF_ERROR(xpress_->PostSolve()) << "Post-solve failed in XPRESS";
+  }
+  // Disable screen output right after solve
+  if (parameters.enable_output()) {
+    RETURN_IF_ERROR(xpress_->SetIntAttr(XPRS_OUTPUTLOG, 0))
+        << "Unable to disable XPRESS logs";
+  }
+  return absl::OkStatus();
+}
+
+absl::Status XpressSolver::SetLpIterLimits(
+    const SolveParametersProto& parameters) {
+  // If the user has set no limits, we still have to reset the limits
+  // explicitly to their default values, else the parameters could be kept
+  // between solves. XPRESS default value for XPRS_LPITERLIMIT is 2147483647
+  int lp_iter_limit = parameters.has_iteration_limit()
+                          ? parameters.iteration_limit()
+                          : 2147483647;
+  // XPRESS default value for XPRS_BARITERLIMIT is 500
+  int bar_iter_limit =
+      parameters.has_iteration_limit() ? parameters.iteration_limit() : 500;
+  RETURN_IF_ERROR(xpress_->SetIntAttr(XPRS_LPITERLIMIT, lp_iter_limit))
+      << "Could not set XPRS_LPITERLIMIT";
+  RETURN_IF_ERROR(xpress_->SetIntAttr(XPRS_BARITERLIMIT, bar_iter_limit))
+      << "Could not set XPRS_BARITERLIMIT";
+  return absl::OkStatus();
+}
+
+absl::StatusOr<SolveResultProto> XpressSolver::ExtractSolveResultProto(
+    absl::Time start, const ModelSolveParametersProto& model_parameters) {
+  SolveResultProto result;
+  ASSIGN_OR_RETURN(SolutionsAndClaims solution_and_claims,
+                   GetSolutions(model_parameters));
+  for (SolutionProto& solution : solution_and_claims.solutions) {
+    *result.add_solutions() = std::move(solution);
+  }
+  ASSIGN_OR_RETURN(*result.mutable_solve_stats(), GetSolveStats(start));
+  ASSIGN_OR_RETURN(const double best_primal_bound, GetBestPrimalBound());
+  ASSIGN_OR_RETURN(const double best_dual_bound, GetBestDualBound());
+  auto solution_claims = solution_and_claims.solution_claims;
+  ASSIGN_OR_RETURN(*result.mutable_termination(),
+                   ConvertTerminationReason(solution_claims, best_primal_bound,
+                                            best_dual_bound));
+  return result;
+}
+
+absl::StatusOr<double> XpressSolver::GetBestPrimalBound() const {
+  return xpress_->GetDoubleAttr(is_mip_ ? XPRS_MIPOBJVAL : XPRS_LPOBJVAL);
+}
+
+absl::StatusOr<double> XpressSolver::GetBestDualBound() const {
+  // TODO: setting LP primal value as best dual bound. Can this be improved?
```

Comment:
There is [XPRS_BESTBOUND](https://www.fico.com/fico-xpress-optimization/docs/latest/solver/optimizer/HTML/BESTBOUND.html) for this.